### PR TITLE
Add external versioner (ref #573)

### PR DESCRIPTION
**Pull request by alex2108**
next try because github does not let me open the old one again :(

### CODE REVIEW

File: internal/versioner/external.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	} else {
+		return errors.New("Versioner: file was not removed by external script")
+	}
+
+	return nil
```
Comment (AudriusButkevicius) [inline internal/versioner/external.go]: this return is redundant.

File: internal/versioner/external.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+	if debug {
+		l.Debugln("archiving", filePath)
+	}
+
+	inFolderPath, err := filepath.Rel(v.folderPath, filePath)
+	if err != nil {
+		return err
+	}
+	if v.command != "" {
```
Comment (AudriusButkevicius) [inline internal/versioner/external.go]: I would do 

```
if v.command == "" { 
  return error
}```

This way the else part losses a level of indentation.

File: internal/versioner/external.go (hunk 1 above)
Comment (alex2108) [inline internal/versioner/external.go]: go1.4 does not seem to complain about that, I should update my version :laughing: , fixed now

File: internal/versioner/external.go (hunk 2 above)
Comment (alex2108) [inline internal/versioner/external.go]: looks much better like this, fixed :)

File: internal/versioner/external.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	// return error if the file was not removed
+	_, err = os.Lstat(filePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	} else {
```
Comment (AudriusButkevicius) [inline internal/versioner/external.go]: `go vet` should complain and suggest dropping the else I think.
But I don't are too much ;)

Comment (calmh) [inline internal/versioner/external.go]:
```go
if _, err = os.Lstat(filePath); os.IsNotExist(err) {
   return nil
}
return errors.New("Versioner: file was not removed by external script")
```

should suffice and cut some code while retaining clarity, but yeah it's only a style issue.

File: internal/versioner/external.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	if v.command == "" {
+		return errors.New("Versioner: command is empty, please enter a valid command")
+	}
+
+	cmd := exec.Command(v.command, v.folderPath, inFolderPath)
+	env := os.Environ()
+	// filter STGUIAUTH and STGUIAPIKEY from environment variables
+	fiteredEnv := []string{}
```
Comment (calmh) [inline internal/versioner/external.go]: "filteredEnv" :)

File: internal/versioner/external.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	_, err := os.Lstat(filePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if debug {
+				l.Debugln("not archiving nonexistent file", filePath)
+			}
+			return nil
+		}
+		return err
+	}
```
Comment (calmh) [inline internal/versioner/external.go]:
```go
if os.IsNotExist(err) {
  if debug {
    l.Debugln("not archiving nonexistent file", filePath)
  }
  return nil
} else if err != nil {
  return err
}
```

reduces nesting

File: internal/versioner/external.go (hunk 4 above)
Comment (alex2108) [inline internal/versioner/external.go]: oops, looks like I copied the variable name with the typo :laughing:

File: internal/versioner/external.go (hunk 5 above)
Comment (alex2108) [inline internal/versioner/external.go]: will update that on all versioner then, I think all of them have exactly that part

File: internal/versioner/external.go (hunk 3 above)
Comment (alex2108) [inline internal/versioner/external.go]: changed that, looks better now :)

File: internal/versioner/external.go (hunk 5 above)
Comment (calmh) [inline internal/versioner/external.go]: :+1:

**Comment by alex2108** [comment]
updated :)

**Comment by AudriusButkevicius** [comment]
LGTM!

**Comment by calmh** [comment]
:+1: Thank for your work and patience!

**Pull request merged**